### ISSUE TITLE
[sklearn#34072] Test infrastructure and documentation to identify unexpected mutation in parallel code

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#356]
**Mirrored from [scikit-learn#34072](https://github.com/scikit-learn/scikit-learn/issues/34072)**

Original title: Test infrastructure and documentation to identify unexpected mutation in parallel code
Original labels: module:test-suite, free-threading
Created: 2026-05-20T20:55:51Z

---

### Describe the workflow you want to enable

To ensure thread-safety, it's useful to have test infrastructure that will provide:

* A way to identify thread-safety issues in code that has previously mostly been run in process pools. Free-threading's lack of GIL makes thread safety issues more likely than in the past.
* A way to prevent _future changes_ to the code introducing new thread-safety problems.
* 🆕 A way to identify which parts of a data structure are intended to be thread-safe, and which are not.

A good starting point here is focusing on operations that are _already_ parallel, where users might reasonably expect to be able to turn threading on as a `joblib` backend and have it Just Work. These operations also work in a very particular way: create some tasks, run them in a worker pool, get back the results of the tasks.

What can go wrong in this particular idiom? Switching from a process pool to a thread pool introduces thread-safety issues with the _inputs_ to a `Parallel()` operation: whereas mutating in a process is fine, mutating in a thread can affect other threads, or future interactions with the object in the main thread.

This suggests a starting point for a invariant to be using in testing: **the inputs to a `Parallel()` operation that can correctly run with both process and thread pools should not change when the function is run.**

🆕 This invariant will not be true for objects that are _designed_ to be mutated safely, so there needs to be a way to opt out of the check.

### Describe your proposed solution

When `pytest` is run (no extra flags! 🆕) and `Parallel()`'s backend a thread pool

* Arguments passed to functions called in parallel are tracked to see if they changed as part of execution.
* "Change" is calculated heuristically, with a hash.
* If hashes change, raise an AssertionError.

🆕 There will be a mechanism to opt-out parameters, or parameter's attributes, from the hashing. In particular, they can be optionally be marked as one of:

* `thread-safe`: indicating these attributes are designed to be safely mutated from multiple threads. This should include a short explanation or pointer to explanation.
* `buggy-thread-unsafe`: indicating this is not thread-safe but is still being accessed from multiple threads. This should include a reference to a GitHub issue explaining the problem.

In addition to opt-in out of the hashing test, these will therefore also serve as documentation of thread-safety or lack thereof. They can also suggest utility classes to formalize certain idioms that repeat, e.g. one can imagine a Rust-style lock that _wraps_ an object, rather than having lock and lock-protected-object being two different parameters. Such classes would then also be automatically opted-out.

When pytest is first run with the hashing, it will fail in three ways; initial tests suggest perhaps 20 locations are affected:

* **Known thread-unsafe code:** For example, BinaryTree. This can be annotated with `ThreadSync`, with 

...(truncated)

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [62](https://github.com/godofecht/flow-scikit/issues/62) in the flow-scikit repository. Further library-specific triage and implementation will happen there.

**Comment by godofecht** [comment]
Moved to [63](https://github.com/godofecht/flow-scikit/issues/63) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
